transient/sshfs.py

```diff
@@ -115,127 +118,131 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-            if is_slow:
-                # We timed out without getting to the 'echo' in the script, even with
-                # extra time. Just give up.
-                raise RuntimeError(f"SSHFS mount timed out: {stderr}")
-            else:
-                logging.warning(
-                    "sshfs mount did not complete. Trying with longer timeout"
-                )
-                # Try again with a longer timeout. The system may just be extremely
-                # slow (like qemu with 1 core and no kvm acceleration)
-                do_sshfs_mount(
-                    connect_timeout=connect_timeout,
-                    local_dir=local_dir,
-                    remote_dir=remote_dir,
-                    local_user=local_user,
-                    local_password=local_password,
-                    ssh_config=ssh_config,
-                    is_slow=True,
+            # Everything from here on out simply verifies that nothing went wrong.
+            if ssh_proc.stderr is not None:
+                first_stderr_line = ssh_proc.stderr.readline().decode("utf-8")
+                if "TRANSIENT_SSH_COMPLETE" not in first_stderr_line:
+                    stderr = first_stderr_line + ssh_proc.communicate()[1].decode("utf-8")
+                    raise RuntimeError(f"SSH connection failed with: {stderr}")
+
+        try:
+            # Now that the SSH connection is established, the SSHFS timeout can be very short
+            stderr = ssh_proc.communicate(timeout=_SSHFS_MAX_RUN_TIME)[1].decode("utf-8")
+        except subprocess.TimeoutExpired:
+            # Because SSHFS is communicating over stdin/out of the ssh connection, SSH
+            # needs to run for as long as the mount exists. Instead of waiting until the
+            # connection is closed, we wait a short time so SSHFS has a chance to fail.
+            # Timing out is expected.
+            pass
+        else:
+            sftp_proc.kill()
+
+            # Check whether this failure looks like it is the result of the vm being
+            # unprovisioned
+            global provision_success
+            if _should_provision(provision_success is True, stderr) is False:
+                raise RuntimeError(f"SSHFS mount failed with: {stderr}")
+
+            with provision_lock:
+                if provision_success is None:
+                    provision_success = provision_system(
+                        self.connect_timeout, self.ssh_config
+                    )
+            if not provision_success:
+                raise RuntimeError(
+                    f"Unable to provision. SSHFS mount failed with: {stderr}"
                 )
+
+            # Retry now that the system is provisioned
+            return self.do_mount()
+
+        # Verify that the server didn't die while communicate() waited for the client
+        assert not sftp_proc.poll()
+
+        # Wake up the main thread
+        self.is_complete.set()
+
+        if ssh_proc.wait():
```

Comment:
Hmm, maybe we want to log the error here as well, as the actual exception may never been seen, right?

Reply:
No. Whenever a Python thread hits an unhandled exception, the output goes to stderr, regardless of whether it's the main thread or a daemon thread. Any time this hits an SSHFS error, it always prints it at the end of a Python traceback.